bioptim/limits/constraints.py

```diff
@@ -299,6 +307,194 @@ def time_constraint(_: Constraint, all_pn: PenaltyNodeList, **unused_param):
+            qddot = nlp.states["qddot"].mx if "qddot" in nlp.states.keys() else nlp.controls["qddot"].mx
+
+            if nlp.external_forces:
+                raise NotImplementedError(
+                    "This implicit constraint tau_equals_inverse_dynamics is not implemented yet with external forces"
+                )
+                # Todo: add fext tau_id = nlp.model.InverseDynamics(q, qdot, qddot, fext).to_mx()
+                # fext need to be a mx
+
+            tau_id = nlp.model.InverseDynamics(q, qdot, qddot).to_mx()
+            var = [nlp.states["q"], nlp.states["qdot"]]
+            dynamic_name = nlp.dynamics_type.type.name
+            if dynamic_name == "TORQUE_DERIVATIVE_DRIVEN":
+                var.extend(
+                    [
+                        nlp.states["tau"],
+                        nlp.controls["taudot"],
+                        nlp.states["qddot"],
+                        nlp.controls["qdddot"],
+                    ]
+                )
+
+            elif dynamic_name == "TORQUE_DRIVEN":
+                var.extend(
+                    [
+                        nlp.controls["tau"],
+                        nlp.controls["qddot"],
+                    ]
+                )
+            else:
+                raise NotImplementedError(
+                    "Implicit Dynamics has not been implemented yet"
+                    f" for such as dynamic: {dynamic_name}. "
+                    "It is only implemented for these dynamics:"
+                    " TORQUE_DRIVEN and TORQUE_DERIVATIVE_DRIVEN."
+                )
+
+            return BiorbdInterface.mx_to_cx("InverseDynamics", tau_id - tau, *var)
+
+        @staticmethod
+        def implicit_soft_contact_forces(_: Constraint, all_pn: PenaltyNodeList, **unused_param):
+            """
+            The time constraint is taken care elsewhere, but must be declared here. This function therefore does nothing
+
+            Parameters
+            ----------
+            _: Constraint
+                The actual constraint to declare
+            all_pn: PenaltyNodeList
+                The penalty node elements
+            **unused_param: dict
+                Since the function does nothing, we can safely ignore any argument
+            """
+
+            nlp = all_pn.nlp
+
+            force_idx = []
+            for i_sc in range(nlp.model.nbSoftContacts()):
+                force_idx.append(3 + (6 * i_sc))
+                force_idx.append(4 + (6 * i_sc))
+                force_idx.append(5 + (6 * i_sc))
+
+            soft_contact_all = nlp.soft_contact_forces_func(nlp.states.mx, nlp.controls.mx, nlp.parameters.mx)
+            soft_contact_force = soft_contact_all[force_idx]
+
+            var = [nlp.states["q"], nlp.states["qdot"]]
+            dynamic_name = nlp.dynamics_type.type.name
+            if nlp.dynamics_type.params["implicit_dynamics"]:
+                if dynamic_name == "TORQUE_DERIVATIVE_DRIVEN":
```

Comment:
Please do not compare (probably ever in your programming career) to str, use the Constraint.TORQUE... and not the name of it. The reason is if we change the name, how do we even know this exists?

Reply:
done